### PR TITLE
Handle OC-Total-Length in new chunking

### LINKED ISSUE #26988: ChunkingNG: Missing chunks are silently ignored
### Steps to reproduce
1. Start a chunked upload with a MKCOL, setting OC-Total-Length to 10M.
2. Upload some, but not all, of the 5 2M chunks.
3. MOVE .file to the destination.
4. The MOVE will succeed. The resulting file will not be 10M in size.

### Expected behaviour
The MOVE should fail because chunks are missing.

### Actual behaviour
The MOVE succeeds and the target file is a concatenation of the existing chunks. 

While not uploading some chunks is a user error, the server should guard against creating corrupted files. Comparison to the OC-Total-Length header allows easy detection of missing data. Additionally, if the server uses integer chunk numbers, holes in the chunk numbering could be detected (e.g. detect that only chunks 0, 1, 3, 4 were uploaded)

### Server configuration
**ownCloud version:** master (6f726844d0164197c83872cd960a179185274ccb)

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Fix short coming of new chunking implementation - 

## Related Issue
fixes #26988

## Motivation and Context
Behave properly ...

## How Has This Been Tested?
Tested with the cli tool as shipped in apps/dav/bin

Please test on various browsers
- [x] chrome
- [x] firefox
- [x] ie
- [ ] edge
- [x] safari

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
@DeepDiver1975 a few weeks ago I was wondering whether X-OC-Mtime on the final move was considered with new chunking. A local test showed that it was considered. But now in this PR you've added extra code for this. I suspect that the reason it worked was maybe that the final MOVE does a `Sabre\File::put()` operation with the data which itself contains the X-OC-Mtime handling. If we leave your PR as is, we'll end up touching the file twice.

We can't easily move away from 'Sabre\File::put()` due to the part file and hooks logic there. Maybe need some kind of flag ?

File: apps/dav/lib/Upload/ChunkingPlugin.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	 * @param string $sourcePath source path
+	 * @param string $destination destination path
+	 */
+	function afterMove($sourcePath, $destination) {
+		if (!$this->sourceNode instanceof FutureFile) {
+			// skip handling as the source is not a chunked FutureFile
+			return;
+		}
+
+		$mTime = $this->server->httpRequest->getHeader('X-OC-mtime');
```
Comment (PVince81) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: no more sanitize?

File: apps/dav/lib/Upload/ChunkingPlugin.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	/**
+	 * @throws BadRequest
+	 */
+	private function verifySize() {
+		$expectedSize = $this->server->httpRequest->getHeader('X-OC-Total-Size');
+		if ($expectedSize === null) {
+			throw new BadRequest('Http header X-OC-Total-Size is missing');
+		}
+		$actualSize = $this->sourceNode->getSize();
+		if ($expectedSize != $actualSize) {
```
Comment (PVince81) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: use strict `!==`

**Review by PVince81 — COMMENTED** [review]
Looks good in general, some minor problem with mtime touching that happens twice.

See comments.

Integration tests will tell whether chunking still works.

Would be good to add integration tests for X-OC-Total-Size.

File: apps/dav/lib/Upload/ChunkingPlugin.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: done in Node::touch

File: apps/dav/lib/Upload/ChunkingPlugin.php (hunk 2 above)
Comment (DeepDiver1975) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: the header is a string and I dont want to cast it to an int due to 32/64 bit nightmare

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 a few weeks ago I was wondering whether X-OC-Mtime on the final move was considered with new chunking. A local test showed that it was considered. But now in this PR you've added extra code for this. I suspect that the reason it worked was maybe that the final MOVE does a Sabre\File::put() operation with the data which itself contains the X-OC-Mtime handling. If we leave your PR as is, we'll end up touching the file twice.

I'll retest this scenario .... THX for the hint

File: apps/dav/lib/Upload/ChunkingPlugin.php (hunk 2 above)
Comment (PVince81) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: what does PHP do internally ? whatever PHP does by itself could be even worse ?

Comment (PVince81) [inline apps/dav/lib/Upload/ChunkingPlugin.php]: if you insist on keeping it, please add a PHP comment to justify this exception for not respecting the coding style

**Comment by DeepDiver1975** [comment]
aaaaargh ... Jenkins was restarted? 💣

**Comment by guruz** [comment]
logic wise the code looks ok to me.
I guess it is ok to mandate X-OC-Total-Size. But then you could also mandate X-OC-Mtime?

**Comment by DeepDiver1975** [comment]
> I haven't tested it though and I don't know why the test fails

tests are failing because X-OC-Total-Size is missing in the tests.

Is X-OC-Total-Size considered mandatory?

**Comment by ogoffart** [comment]
Hold on!  I just checked the client source and the header is called `OC-Total-Lenght` (No X-, and Lenght).
Also, we only sent this header from MOVE since the client 2.3.2,  so the version 2.3.0 and 2.3.1 would break if we make the header mandatory. Hence it should not be mendatory.

**Comment by DeepDiver1975** [comment]
> Hold on! I just checked the client source and the header is called OC-Total-Lenght (No X-, and Lenght).

the server code in the old chunking used OC-Total-Length - are you sure about the twisted chars?
Can we correct this? Kind of *peinlich* ....

**Comment by ogoffart** [comment]
Yes, we re-used the same header on purpose.

**Comment by DeepDiver1975** [comment]
jenkins green - please test @SamuAlfageme @ogoffart @guruz

**Comment by guruz** [comment]
We even have the wrong changelog :-)
Size vs Length

```
guruz@mgmacbookpro mirall $ git grep -i 'Total-'
ChangeLog:* Improve compatibility with server 10.0 (#5691, X-OC-Total-Size)
csync/ChangeLog:  * Add OC-Total-Length header for better quota handling.
src/libsync/propagateuploadng.cpp:    headers["OC-Total-Length"] = QByteArray::number(_item->_size);
src/libsync/propagateuploadng.cpp:        headers["OC-Total-Length"] = QByteArray::number(fileSize);
src/libsync/propagateuploadv1.cpp:    headers["OC-Total-Length"] = QByteArray::number(fileSize);
```

**Comment by mrow4a** [comment]
@DeepDiver1975 will have a look

**Comment by guruz** [comment]
@mrow4a To clarify: The wrong size message needs to be displayed, not the "The computed checksum does not match the one received from the client.".

(So actually this issue is not sooo important as the client sends a checksum anyway ;-) )

**Comment by PVince81** [comment]
restarted CI job

**Comment by PVince81** [comment]
Would be good to adjust the integration tests about new chunking too to also send this header

**Comment by PVince81** [comment]
Integration tests to add:

- send the wrong total size and verify that the upload failed
- send the correct total size and verify that the upload succeeded

**Comment by DeepDiver1975** [comment]
tested with Chromium Version 60.0.3112.78

**Comment by DeepDiver1975** [comment]
tested with firefox 53.0 (64-bit)

File: apps/files/js/file-upload.js [hunk 3]
```diff
@@ -284,11 +284,22 @@ OC.FileUpload.prototype = {
 		}
 
 		var uid = OC.getCurrentUser().uid;
+		var mtime = this.getFile().lastModified;
+		var size = this.getFile().size;
+		var headers = {};
+		if (mtime) {
+			headers['X-OC-Mtime'] = mtime / 1000;
+		}
+		if (size) {
```
Comment (PVince81) [inline apps/files/js/file-upload.js]: This would exclude the size "0" but then the chunking code isn't triggered for such sizes, so it's ok

**Review by PVince81 — APPROVED** [review]
Code changes look fine 👍

**Comment by individual-it** [comment]
tested with IE & Safari both are sending "OC-Total-Length" coirrectly

**Comment by PVince81** [comment]
well, and also fixes a critical bug when you upload a file with new chunking in IE11 which results in EPOCH being set as mtime

**Comment by DeepDiver1975** [comment]
what about edge? does it work there as well? @phil-davis @individual-it THX

**Comment by phil-davis** [comment]
Edge on Win10 laptop with current August "patch Tuesday" patches all applied:
oC server on Ubuntu VM PHP dev server running core master:
```
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: en-AU, en-US; q=0.8, en; q=0.6, ne-NP; q=0.4, ne; q=0.2
Cache-Control: no-cache
Connection: Keep-Alive
Content-Length: 0
Cookie: ocr6rsb60140=9m02hs1jeq22592nvbju4438j1; oc_sessionPassphrase=ba1k5GOeSzLVJVIU%2BS6OZ31lCrpxrgPSjPqHM%2F469yMMlJBSZm3xaYdbzgsuvREXO%2BzJyDOtk1n18U%2BZHPGiG4%2FwajB8tMB9u%2FmG3VXHkVXRT7xMs4qpLaIWbTZGNX6H
Destination: http://10.49.209.23:8080/remote.php/dav/files/admin/LION_Title_01_01.mp4
Host: 10.49.209.23:8080
requesttoken: ZhcOLwMYHAJnFhgJYHANTQA5EFd8WwQWfEULdS5QPmA=:MdDafpOkHr7x9CObkiu9LbJs1umGvaHS9jRzHh6Teac=
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393
X-OC-Mtime: NaN
X-Requested-With: XMLHttpRequest
```

oC server on Ubuntu VM PHP dev server running core verify-chunking-total-size branch (this code):
```
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: en-AU, en-US; q=0.8, en; q=0.6, ne-NP; q=0.4, ne; q=0.2
Cache-Control: no-cache
Connection: Keep-Alive
Content-Length: 0
Cookie: ocr6rsb60140=9m02hs1jeq22592nvbju4438j1; oc_sessionPassphrase=ba1k5GOeSzLVJVIU%2BS6OZ31lCrpxrgPSjPqHM%2F469yMMlJBSZm3xaYdbzgsuvREXO%2BzJyDOtk1n18U%2BZHPGiG4%2FwajB8tMB9u%2FmG3VXHkVXRT7xMs4qpLaIWbTZGNX6H
Destination: http://10.49.209.23:8080/remote.php/dav/files/admin/LION_Title_01_01.mp4
Host: 10.49.209.23:8080
OC-Total-Length: 1443516495
requesttoken: fRcSIVFYYwJoJW0wMmVyFi0lPSd8XDcsO1wHQioCJAA=:VdXo400kGABAkV09FuXILeyIvlapr3R3gZ2u+eybLxE=
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393
X-Requested-With: XMLHttpRequest
```

``OC-Total-Length`` is good - the size of the movie uploaded.
``X-OC-Mtime: NaN`` has disappeared - without this PR the file gets 1 Jan 1970 date, with this PR the file gets the current date-time??? But when I do it in Chrome on Win10 (both before and with the PR) it sends ``X-OC-Mtime: 1487326866``

**Comment by PVince81** [comment]
So this is ready for merge after CI ?

@DeepDiver1975 can you prepare the stable10 PR in advance ? Thanks

**Comment by DeepDiver1975** [comment]
> So this is ready for merge after CI ?

yes
 
> @DeepDiver1975 can you prepare the stable10 PR in advance ? Thanks

done

**Pull request merged**